src/probability/extrealScript.sml

```diff
@@ -9463,6 +9463,268 @@ Proof
+        suffices_by rw[EXTREAL_SUM_IMAGE_ALT_FOLDR] >>
+    Induct_on ‘l’ >> rw[listTheory.FOLDR] >> irule le_add2 >> simp[]
+QED
+
+Theorem EXTREAL_SUM_IMAGE_COUNT_ZERO:
+    ∀f. ∑ f (count 0) = 0:extreal
+Proof
+    simp[COUNT_ZERO,EXTREAL_SUM_IMAGE_EMPTY]
+QED
+
+Theorem EXTREAL_SUM_IMAGE_COUNT_ONE:
+    ∀f. ∑ f (count 1) = f 0:extreal
+Proof
+    simp[COUNT_ONE,EXTREAL_SUM_IMAGE_SING]
+QED
+
+Theorem EXTREAL_SUM_IMAGE_COUNT_SUC:
+    ∀f n. (∀m. m ≤ n ⇒ f m ≠ −∞) ∨ (∀m. m ≤ n ⇒ f m ≠ +∞) ⇒
+        ∑ f (count (SUC n)) = (∑ f (count n)) + f n:extreal
+Proof
+    rw[] >> ‘count (SUC n) = (count n) ∪ {n}’ by fs[count_def,EXTENSION] >>
+    rw[] >> pop_assum kall_tac >>
+    ‘∑ f (count n ∪ {n}) = ∑ f (count n) + ∑ f {n}’ suffices_by fs[EXTREAL_SUM_IMAGE_SING] >>
+    irule EXTREAL_SUM_IMAGE_DISJOINT_UNION >> simp[]
+QED
+
+(*** EXTREAL_PROD_IMAGE Theorems ***)
+
+Theorem EXTREAL_PROD_IMAGE_NOT_INFTY:
+    ∀f s. FINITE s ∧ (∀x. x ∈ s ⇒ f x ≠ −∞ ∧ f x ≠ +∞) ⇒ ∏ f s ≠ −∞ ∧ ∏ f s ≠ +∞
+Proof
+    strip_tac >> simp[Once $ GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >> CONJ_TAC
+    >- simp[EXTREAL_PROD_IMAGE_EMPTY,SYM normal_1] >>
+    NTAC 5 strip_tac >> fs[EXTREAL_PROD_IMAGE_PROPERTY,DELETE_NON_ELEMENT_RWT] >>
+    Cases_on ‘f e’ >> Cases_on ‘∏ f s’ >> rfs[extreal_mul_def]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_NORMAL:
+    ∀f s. FINITE s ⇒ ∏ (λx. Normal (f x)) s = Normal (∏ f s)
+Proof
+    strip_tac >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,REAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT,extreal_mul_def,normal_1]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_0:
+    ∀f s. FINITE s ∧ (∃x. x ∈ s ∧ f x = 0) ⇒ ∏ f s = 0
+Proof
+    NTAC 2 strip_tac >> simp[GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT] >- fs[] >>
+    DISJ2_TAC >> first_x_assum irule >> qexists_tac ‘x’ >> simp[]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_1:
+    ∀f s. FINITE s ∧ (∀x. x ∈ s ⇒ f x = 1) ⇒ ∏ f s = 1
+Proof
+    NTAC 2 strip_tac >> simp[GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_ONE:
+    ∀s. FINITE s ⇒ ∏ (λx. 1) s = 1x
+Proof
+    Induct_on ‘s’ >> simp[EXTREAL_PROD_IMAGE_EMPTY,EXTREAL_PROD_IMAGE_PROPERTY,DELETE_NON_ELEMENT_RWT]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_POS:
+    ∀f s. FINITE s ∧ (∀x. x ∈ s ⇒ 0 ≤ f x) ⇒ 0 ≤ ∏ f s
+Proof
+    strip_tac >> simp[GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT] >> irule le_mul >> simp[]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_MONO:
+    ∀f g s. FINITE s ∧ (∀x. x ∈ s ⇒ 0 ≤ f x ∧ f x ≤ g x) ⇒ ∏ f s ≤ ∏ g s
+Proof
+    NTAC 2 strip_tac >> simp[GSYM AND_IMP_INTRO] >> Induct_on ‘s’ >>
+    rw[EXTREAL_PROD_IMAGE_THM,DELETE_NON_ELEMENT_RWT] >> irule le_mul2 >>
+    simp[EXTREAL_PROD_IMAGE_POS]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_COUNT_ZERO:
+    ∀f. ∏ f (count 0) = 1x
+Proof
+    simp[COUNT_ZERO,EXTREAL_PROD_IMAGE_EMPTY]
+QED
+
+Theorem EXTREAL_PROD_IMAGE_COUNT_ONE:
```

Comment:
`[simp]` (unless there is already a simp for `PI f {x}`; and if there isn't, why isn't there?)

Reply:
This raises questions about:
`COUNT_ONE`
`REAL_PROD_IMAGE_SING` (doesn't exist, I could make it)
`EXTREAL_PROD_IMAGE_SING` (the thing you ask about existing)
`REAL_SUM_IMAGE_SING`
`EXTREAL_SUM_IMAGE_SING`
`EXTREAL_SUM_IMAGE_COUNT_ONE`

As for why `EXTREAL_PROD_IMAGE_SING` isn't a `[simp]`, that calculus predates me. I suspect whoever initially proved it either didn't know about `[simp]`s, or was worried about making too many things `[simp]`s (although perhaps I am projecting).

That said, I'm happy to make whatever `[simp]`s. Incidentally, how do you do that with the old `store_thm()` syntax?

Reply:
In old `store_thm()` syntax you can do the same: just put `[simp]` as part of the theorem name string (only with that string but SML variable name). The older way is to call `export_rewrites`, in `pred_setScript.sml` there should be such examples.

I personally tend to not declare too much `[simp]` for less common logical constants. Initially I consider `EXTREAL_PROD_IMAGE` is only needed in very few cases, and I should bother too much the automatic simplifier.  Furthermore, sometimes I meet the cases that a goal is simplified too much due to some later-added `[simp]` and as a solution I may remove some unwanted `[simp]`.

I suggest for now you can just follow what @mn200 suggests and get the PR merged ASAP. Later in proving other theorems it is always possible to adjust set `[simp]` properties (before next official release).

Reply:
I absolutely intend to follow Michael's suggestions, but I just also wanted to solicit his judgement on other similar theorems while I was at it.

But I also want to get the PR to an acceptable state ASAP. Especially since this is only the first batch of results. I wanted to do the Lebesgue stuff in a separate batch, because I suspected there would be some issues. I'll start working on isolating the stuff tomorrow.

Reply:
Can confirm, all of these break things. So, @mn200, what are the magical incantations to disable and re-enable simps? I'm not seeing anything obvious in simpLib.

Reply:
Use `Excl "thmname"` as an argument in the list of theorems passed to `simp` and friends.   Or, to induce a more long-lasting change, use `val _ = temp_delsimps ["thmname1", "thmname2"]`.

Reply:
The `temp_delsimps` is more what I was looking for. Anything to bring a simp back later on after that? (Basically allowing me to disable them for a number of proofs within the script.)

Reply:
Yes (`augment_srw_ss`), but it's super gross to leave a script in a state where non-obvious chunks of it will behave differently because of theorems jumping in and out of the global simpset value.  If you're going to use `temp_delsimps` put it at the top of the file and don't undo it later on.

Reply:
I think, if one has to prove something by excluding some theorems instead of applying some, this would be counter-intuitive to what humans usually do.

Reply:
Done